### ISSUE TITLE
Missing a cast to abstract type for nested fields of object declaration

### ISSUE THREAD

**Issue by RealyUniqueName** [HaxeFoundation/haxe#6003]
In the following code i expect `{field:'world'}` to be also casted to `Struct`:
```haxe
class Test {
    static function main() {
        check({ nested: {field:'world'} });
        check(get());
    }

    static function check(o:Struct<Options>) {}

    static function get():Options {
        return { nested: {field:'world'} };
    }
}

abstract Struct<T:{}>(String) from String to String {
    @:from
    static public function fromObject<T:{}>(v:T):Struct<T> return Std.string(v);
}

typedef Options = {?nested: Struct<{?field:String}>} //also if i make `nested` a mandatory field it does not even compile
```
But compiler does not generate implicit cast for the first case. Here is dump=pretty:
```haxe
@:directlyUsed @:used
class Test {

	@:keep
	static function main() {
		Test.check(@:implicitCast _Test.Struct_Impl_.fromObject({nested : {field : "world"}}));
		Test.check(@:implicitCast _Test.Struct_Impl_.fromObject(Test.get()));
	}

	static function check(o:Struct<Options>) {}

	static function get() {
		return {nested : @:implicitCast _Test.Struct_Impl_.fromObject({field : "world"})};
	}
}
```

**Comment by nadako** [comment]
This is a regression from 3.2.1. It doesn't generate implicit cast there too, but at least it emits a compile error stead of silently generating wrong code: http://try-haxe.mrcdk.com/#d673B

**Comment by Simn** [comment]
Too risky to fix this for  3.4, regression or not. Structural typing is a mess anyway.

**Comment by ousado** [comment]
TL;DR I think the behavior is correct now and was buggy in 3.2.1

I don't think it should generate the nested fields cast in the first case, the reasoning is this:
The structure is unified with `Struct<Options>`, hence `@:from` is called, the structure has to unify with the more general `T:{}`, but not `Options`. Then a `String` is returned, which unifies with `Struct<T>`, which from the callsite's perspective happens to be `Struct<Options>`, but the value is passed to Std.string which takes a `Dynamic`, so again no need for it to unify with `Options` directly. 

Because there's no point at which the compiler needs to unify any value's type with `Options`, the nested field which typed as `Struct<{?field:String}>` never becomes relevant for the conversion of an actual value.

**Comment by RealyUniqueName** [comment]
`Std.string` and `String` don't matter in this example. In my original case it was an extern type.
Also did you notice that in `get()` function everything is correctly casted as expected?

**Comment by ousado** [comment]
@RealyUniqueName yes, the difference is that the return type of that function is `Options`, so the conversion is triggered there. 

I removed the optional fields and added a few more cases here: http://try.haxe.org/#fFE31 which show that there are limits even when the type is made explicit: `var v2 = (v:Options);` or the v3 case aren't possible either, which makes sense, since there's a reference around and the type of an existing field would be modified in place, otherwise. 

But when there's an object literal typed as Options and no reference to the struct is around like in `check( ({nested:{field:""}}:Options) )` and in the return of your function, the conversion can be applied.
